Created API which fetches data from public API and has two methods

Someboty opened the pull request:
Created API which fetches data from public API and has two methods:
first gets a random character in the universe the animated series Rick & Morty; 
second takes a string as an argument, and returns a list of all characters whose name contains the search string.

Someboty committed:
Created API which fetches data from public API and has two methods:
first gets a random character in the universe the animated series Rick & Morty;
second takes a string as an argument, and returns a list of all characters whose name contains the search string.

vovasalyha commented on src/main/java/mate/academy/rickandmorty/service/impl/CharacterClient.java:
```diff
@@ -0,0 +1,60 @@
+        }
+
+        for (int i = 2; i <= pages; i++) {
+            request = HttpRequest.newBuilder()
```
[Optional] If you want to practice concurrency, you can spin up few threads to send requests in parallel (or even better, use CompletableFuture, or even Spring's WebClient which is by default reactive(asynchronous)). [Here's the link](https://www.baeldung.com/spring-5-webclient)

vovasalyha commented on src/main/java/mate/academy/rickandmorty/service/impl/CharacterClientService.java:
```diff
@@ -0,0 +1,22 @@
+    public void saveCharactersToDb() {
+        characterClient.getCharacters().stream()
+                .map(characterMapper::toEntity)
+                .forEach(characterRepository::save);
```
Find out if this approach uses batch inserts by default, if no, then consider using saveAll() method

vovasalyha reviewed (approved):
GJ. Left few thoughts for improvement

Someboty committed:
Added multithreading support in CharacterClient
Made saving characters using batch inserts in CharacterClientService

vovasalyha commented on src/main/java/mate/academy/rickandmorty/service/impl/CharacterClient.java:
```diff
@@ -0,0 +1,72 @@
+        ExecutorService executorService = Executors.newFixedThreadPool(pages);
+        List<CompletableFuture<List<RequestResultsDto>>> futureRequests = new ArrayList<>();
+        for (int i = 2; i <= pages; i++) {
+            Future<List<RequestResultsDto>> future = executorService.submit(
```
Your solution is mostly correct, but there is one important thing to keep in mind. You whether use ExecutorService or CompletableFuture. Using them both makes no sense, because using of executor service already runs your tasks in parallel. So you need to choose one of two options here:
- Use ExecutorService and collect List of Future then process results in a stream
```java
Future<List<RequestResultsDto>> future = executorService.submit(
            new CustomThread(httpClient, objectMapper, i));
    futureRequests.add(future);
```
```java
List<RequestResultsDto> additionalCharacters = futureRequests.stream()
        .map(future -> {
            try {
                return future.get();
            } catch (Exception e) {
                throw new RuntimeException(e);
            }
        })
        .flatMap(List::stream)
        .toList();
```
- Use CompletableFuture to submit tasks asynchronously (CompletableFuture uses embedded ForkJoinPool by default, which is created for each Java app)

```java
CompletableFuture.supplyAsync(() -> new CustomThread(httpClient, objectMapper, i).call())
...
```

But the general idea is correct, GJ.

Someboty committed:
Fixed creating list of CompletableFuture by removing Future creating
Separated initial page's index in separate constant

andrii-hoienko commented on src/main/java/mate/academy/rickandmorty/service/impl/CharacterClient.java:
```diff
@@ -0,0 +1,62 @@
+            throw new RuntimeException(e);
+        }
+        List<CompletableFuture<List<RequestResultsDto>>> futureRequests = new ArrayList<>();
+        for (int i = INITIAL_PAGE + 1; i <= pages; i++) {
```
is it possible to use in IntStream.range (...

andrii-hoienko replied:
I would prefer to add logic of receiving additionalCharacters to separate method

andrii-hoienko commented on src/main/java/mate/academy/rickandmorty/thread/CustomThread.java:
```diff
@@ -0,0 +1,37 @@
+import mate.academy.rickandmorty.dto.external.RequestResultsDto;
+
+@RequiredArgsConstructor
+public class CustomThread implements Callable<List<RequestResultsDto>> {
```
Not representative name

andrii-hoienko commented on src/main/java/mate/academy/rickandmorty/thread/CustomThread.java:
```diff
@@ -0,0 +1,37 @@
+    private final int pageNumber;
+
+    @Override
+    public List<RequestResultsDto> call() {
```
code duplication

andrii-hoienko reviewed (changes requested):
GJ! Let`s improve your solution

Someboty committed:
Separated logic getCharacters into 2 separated methods
Renamed CustomThread into GetRequestDtoThread
Replaced for cycle by IntStream.range()

andrii-hoienko reviewed (approved):
GJ!
